taichi/common/exceptions.h

```diff
@@ -1,38 +1,135 @@
   explicit TaichiExceptionImpl(const std::string msg) : msg_(msg) {
   }
   const char *what() const noexcept override {
     return msg_.c_str();
   }
 };
 
-class TaichiTypeError : public TaichiExceptionImpl {
+class TaichiError : public TaichiExceptionImpl {
   using TaichiExceptionImpl::TaichiExceptionImpl;
 };
 
-class TaichiSyntaxError : public TaichiExceptionImpl {
+class TaichiWarning : public TaichiExceptionImpl {
   using TaichiExceptionImpl::TaichiExceptionImpl;
+
+ protected:
+  static constexpr std::string_view name_;
+
+ public:
+  void emit() {
+    taichi::Logger::get_instance().warn(std::string(name_) + "\n" + msg_);
+  }
 };
 
-class TaichiIndexError : public TaichiExceptionImpl {
-  using TaichiExceptionImpl::TaichiExceptionImpl;
+class TaichiTypeError : public TaichiError {
+  using TaichiError::TaichiError;
 };
 
-class TaichiRuntimeError : public TaichiExceptionImpl {
-  using TaichiExceptionImpl::TaichiExceptionImpl;
+class TaichiSyntaxError : public TaichiError {
+  using TaichiError::TaichiError;
 };
 
-class TaichiAssertionError : public TaichiExceptionImpl {
-  using TaichiExceptionImpl::TaichiExceptionImpl;
+class TaichiIndexError : public TaichiError {
+  using TaichiError::TaichiError;
+};
+
+class TaichiRuntimeError : public TaichiError {
+  using TaichiError::TaichiError;
+};
+
+class TaichiAssertionError : public TaichiError {
+  using TaichiError::TaichiError;
+};
+
+class TaichiIrError : public TaichiError {
+  using TaichiError::TaichiError;
+};
+
+class TaichiCastWarning : public TaichiWarning {
+  using TaichiWarning::TaichiWarning;
+  static constexpr std::string_view name_ = "TaichiCastWarning";
+};
+
+class TaichiTypeWarning : public TaichiWarning {
+  using TaichiWarning::TaichiWarning;
+  static constexpr std::string_view name_ = "TaichiTypeWarning";
+};
+
+class TaichiIrWarning : public TaichiWarning {
+  using TaichiWarning::TaichiWarning;
+  static constexpr std::string_view name_ = "TaichiIrWarning";
+};
+
+class TaichiIndexWarning : public TaichiWarning {
+  using TaichiWarning::TaichiWarning;
+  static constexpr std::string_view name_ = "TaichiIndexWarning";
+};
+
+class TaichiRuntimeWarning : public TaichiWarning {
+  using TaichiWarning::TaichiWarning;
+  static constexpr std::string_view name_ = "TaichiRuntimeWarning";
+};
+
+struct ErrorEmitter {
+  ErrorEmitter() = delete;
+  ErrorEmitter(ErrorEmitter &) = delete;
+  ErrorEmitter(ErrorEmitter &&) = delete;
+
+  // Emit an error on stmt with error message
+  template <typename E,
+            typename = std::enable_if_t<
+                std::is_base_of_v<TaichiExceptionImpl, std::decay_t<E>>>,
+            typename T,
+            typename = std::enable_if_t<
+                std::is_same_v<std::decay_t<decltype(std::declval<T>()->tb)>,
+                               std::string>>>
+  ErrorEmitter(E &&error, T p_stmt, std::string &&error_msg) {
+    if constexpr ((std::is_same_v<std::decay_t<T>, DebugInfo *> ||
```

Comment:
Is `T` of type `Stmt*`? In that case `std::is_same_v<std::decay_t<T>, DebugInfo *>` will never get evaluated to true?

Reply:
`T` could be a pointer to `Stmt`, `Expression`, or `DebugInfo`, as long as the object be pointed to has a member function called `get_tb()` which returns a `std::string`. The name `p_stmt` is misleading and can be changed to something like `p_dbg_info`.
The `if constexpr` here is a workaround to stop printing the same traceback twice, which is subject to change.